internal/datacoord/copy_segment_checker.go

```diff
@@ -0,0 +1,698 @@
+
+	// Report metrics
+	metrics.CopySegmentTasks.WithLabelValues(datapb.CopySegmentTaskState_CopySegmentTaskPending.String()).Set(float64(pending))
+	metrics.CopySegmentTasks.WithLabelValues(datapb.CopySegmentTaskState_CopySegmentTaskInProgress.String()).Set(float64(inProgress))
+	metrics.CopySegmentTasks.WithLabelValues(datapb.CopySegmentTaskState_CopySegmentTaskCompleted.String()).Set(float64(completed))
+	metrics.CopySegmentTasks.WithLabelValues(datapb.CopySegmentTaskState_CopySegmentTaskFailed.String()).Set(float64(failed))
+}
+
+// ============================================================================
+// State Machine: Pending → Executing
+// ============================================================================
+
+// checkPendingJob transitions job from Pending to Executing by creating tasks.
+//
+// This is the first state transition in the job lifecycle. It groups segment ID
+// mappings into tasks (to avoid tasks that are too large) and creates task metadata.
+// The actual file copying is triggered later by the inspector component.
+//
+// Process flow:
+//  1. Check if tasks already exist (idempotent - don't create duplicates)
+//  2. Validate job has segment mappings (empty jobs are marked completed)
+//  3. Split mappings into groups (max segments per task configurable)
+//  4. For each group:
+//     a. Allocate task ID
+//     b. Create task metadata with lightweight ID mappings
+//     c. Save task to metadata store
+//  5. Update job state to Executing with initial progress (0/total)
+//
+// Task grouping:
+//   - Controlled by MaxSegmentsPerCopyTask parameter
+//   - Prevents tasks from becoming too large and timing out
+//   - Enables parallel execution across multiple DataNodes
+//
+// Why lightweight ID mappings:
+//   - Task metadata only stores source→target segment ID mappings
+//   - Full segment metadata (binlogs, indexes) is fetched by DataNode when executing
+//   - Keeps task metadata small and efficient to persist
+//
+// Idempotency:
+//   - Safe to call multiple times - only creates tasks on first call
+//   - Subsequent calls return early if tasks already exist
+func (c *copySegmentChecker) checkPendingJob(job CopySegmentJob) {
+	log := log.With(zap.Int64("jobID", job.GetJobId()))
+
+	// Step 1: Check if tasks already created (idempotent operation)
+	tasks := c.copyMeta.GetTaskBy(c.ctx, WithCopyTaskJob(job.GetJobId()))
+	if len(tasks) > 0 {
+		return
+	}
+
+	// Step 2: Validate job has segment mappings
+	idMappings := job.GetIdMappings()
+	if len(idMappings) == 0 {
+		log.Warn("no id mappings to copy, mark job as completed")
+		c.copyMeta.UpdateJob(c.ctx, job.GetJobId(),
+			UpdateCopyJobState(datapb.CopySegmentJobState_CopySegmentJobCompleted),
+			UpdateCopyJobReason("no segments to copy"))
+		return
+	}
+
+	// Step 3: Split mappings into groups (max segments per task)
+	maxSegmentsPerTask := Params.DataCoordCfg.MaxSegmentsPerCopyTask.GetAsInt()
```

Comment:
is there a reason not use 1 segment as a task?

Reply:
consider two reason: 
1. large collection may has thousands of segment, it causes too much tasks
2. one singe segment copy should be fast, one segment is too small for a task, cause much more workload on schedule progress